### PR TITLE
Tutorial for backend with SQL [16577]

### PR DESCRIPTION
**Pull request by jparisu**
Signed-off-by: jparisu <javierparis@eprosima.com>

### CODE REVIEW

File: docs/rst/tutorials/tools/monitor_sql/monitor_sql.rst [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+
+CMakeLists.txt
+^^^^^^^^^^^^^^
+
+Include at the end of the `CMakeList.txt` file you created earlier the following code snippet.
+This adds all the source files needed to build the executable, and links the executable and the library together.
+
+.. literalinclude:: /../code/monitor_sql_tutorial/CMakeLists.txt
+    :language: cmake
+    :lines: 40-49
```
Comment (juanlofer-eprosima) [inline docs/rst/tutorials/tools/monitor_sql/monitor_sql.rst:181]: Wrong end?

File: docs/rst/tutorials/tools/monitor_sql/monitor_sql.rst [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+At this point the project is ready for building, compiling and running the application.
+From the base workspace directory (`ros_ws`), run the following commands.
+
+.. code-block:: bash
+
+    colcon build
+    source install/setup.bash
+    ros2 run monitor_sql_tutorial monitor_sql_tutorial
+
+Then open two more terminals and load the Vulcanexus environment.
```
Comment (juanlofer-eprosima) [inline docs/rst/tutorials/tools/monitor_sql/monitor_sql.rst:203]:
```suggestion
Now open two more terminals and load the Vulcanexus environment.
```

File: docs/rst/tutorials/tools/monitor_sql/monitor_sql.rst [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+Then, in one of them run a ``talker`` and in the other one a ``listener`` of the ``demo_nodes_cpp`` ROS 2 package, available in the Vulcanexus Desktop distribution.
```
Comment (juanlofer-eprosima) [inline docs/rst/tutorials/tools/monitor_sql/monitor_sql.rst:204]:
```suggestion
Then, run a ``talker`` and a ``listener`` of the ``demo_nodes_cpp`` ROS 2 package, available in the Vulcanexus Desktop distribution, in a different terminal each.
```

File: docs/rst/tutorials/tools/monitor_sql/monitor_sql.rst [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+.. note::
+
+    In order to monitor other *statistics topics*, add them to environment variable :code:`FASTDDS_STATISTICS`.
+    Check the *statistics topics* available in the `Fast DDS Documentation <https://fast-dds.docs.eprosima.com/en/latest/fastdds/statistics/dds_layer/topic_names.html#statistics-topic-names>`_.
+
+Next steps
+----------
+
+Now you can develop more functionalities in your application, such as collecting more performance data or monitoring other topics.
+You can check also :ref:`this tutorial <tutorials_tools_prometheus>` explaining how to connect an application developed with the *Fast DDS Statistics Backend* to a visualization tool like *Grafana*.
```
Comment (juanlofer-eprosima) [inline docs/rst/tutorials/tools/monitor_sql/monitor_sql.rst:231]:
```suggestion
You may also check :ref:`this tutorial <tutorials_tools_prometheus>` explaining how to connect an application developed with the *Fast DDS Statistics Backend* to a visualization tool like *Grafana*.
```

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]

File: docs/rst/tutorials/tools/monitor_sql/monitor_sql.rst (hunk 1 above)
Comment (juanlofer-eprosima) [reply docs/rst/tutorials/tools/monitor_sql/monitor_sql.rst:181]: Also, typo *CMakeList.txt*

File: docs/rst/tutorials/tools/monitor_sql/monitor_sql.rst [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+--------------------------------------------
+
+This section explains the source code required to implement this tutorial.
+However, some code is reused from :ref:`previous tutorial <tutorials_tools_monitor_with_backend>` and will not be repeated here.
+
+
+Creating the application workspace
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The application workspace will have the following structure at the end of the project.
```
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply docs/rst/tutorials/tools/monitor_sql/monitor_sql.rst:78]: NIT: an include folder is also created in the workspace.

File: docs/rst/tutorials/tools/monitor_sql/monitor_sql.rst [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+* :ref:`linux_binary_installation`
+* :ref:`linux_source_installation`
+* :ref:`docker_installation`
+
+Ensure that the Vulcanexus installation includes Vulcanexus Tools (either ``vulcanexus-humble-desktop``, ``vulcanexus-humble-tools``, or ``vulcanexus-humble-base``).
+
+.. note::
+
+    This tutorial uses SQLite3 as SQL library to connect with an in-disk database.
+    This SQLite3 is already installed in ROS2 environment.
```
[the remark this answers is not in the document]
Comment (jparisu) [reply docs/rst/tutorials/tools/monitor_sql/monitor_sql.rst:39]: sure about `the`? Vulcanexus :+1:

File: docs/rst/tutorials/tools/monitor_sql/monitor_sql.rst [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+This tutorial executable :code:`monitor_sql_tutorial` produces a database stored with name :code:`vulcanexus_monitor.db` in the workspace where executed.
+This database contains one table called :code:`data` with 3 columns:
+
+* **timestamp [key]** *milliseconds since epoch* Time since epoch in milliseconds
+* **latency_median** *nanoseconds* Median of Latency in the interval :code:`timestamp - 5000 : timestamp`
+* **throughput_median** *MB/second* Median of Throughput in the interval :code:`timestamp - 5000 : timestamp`
+
+Each 5000 ms the program calls the *Statistics Backend* API and store the results for latency median and
+throughput mean for all Nodes using :code:`chatter` topic.
+The :code:`timestamp` column is the **key** of the table as it cannot be repeated.
```
[the remark this answers is not in the document]
Comment (jparisu) [reply docs/rst/tutorials/tools/monitor_sql/monitor_sql.rst:54]: Indeed, it is only for this tutorial. I think the SQL details must be designed by the final user

File: docs/rst/tutorials/tools/monitor_sql/monitor_sql.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (jparisu) [reply docs/rst/tutorials/tools/monitor_sql/monitor_sql.rst:78]: Yes, but it is useles... I add a comment about it in the proper command

File: code/monitor_sql_tutorial/src/sql_monitor.cpp [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+            now - std::chrono::milliseconds(INTERVAL_MS),   // t_from
+            now,                                            // t_to
+            StatisticKind::MEDIAN);                         // Statistic
+
+        // There is only one value, check it is not nan
+        return data[0];
+    }
+
+    //! Get publication thougput mean
+    StatisticsData get_publication_throughput_mean()
```
[the remark this answers is not in the document]
Comment (jparisu) [reply code/monitor_sql_tutorial/src/sql_monitor.cpp:150]: Latency median and Throughput mean

File: docs/rst/tutorials/tools/monitor_sql/monitor_sql.rst [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+
+Background
+----------
+
+Vulcanexus integrates :ref:`fastdds_statistics_backend`, which is a useful tool for monitoring and studying a ROS 2 network since ROS 2 relies on the `DDS specification <https://www.omg.org/spec/DDS/1.4/About-DDS/>`_ to communicate the different nodes.
+The interaction with this tool is through a C++ API.
+This could be done to create powerful monitoring tools specialized by the user.
+In this tutorial we show how to create an application of a :ref:`fastdds_statistics_backend` connected with a SQL in-disk database.
+
+.. figure:: /rst/figures/tutorials/tools/sql_monitor_backend/summary.png
```
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply docs/rst/tutorials/tools/monitor_sql/monitor_sql.rst:18]: Oh sorry, I was referring to image `database.png`.

File: docs/rst/tutorials/tools/monitor_sql/monitor_sql.rst (hunk 1 above)
Comment (juanlofer-eprosima) [reply docs/rst/tutorials/tools/monitor_sql/monitor_sql.rst]: Typo still there.

**Review by juanlofer-eprosima — APPROVED** [review]
LGTM

**Pull request merged**